In [1]:
#import dependencies
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
%matplotlib inline

In [2]:
#load the data set using pandas
heart_dataset = pd.read_csv('heart_disease_data.csv')

In [3]:
heart_dataset.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [4]:
heart_dataset.tail()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0

In [5]:
heart_dataset.shape

(303, 14)

## So in this data set there are 303 rows with 13 data columns

In [6]:
heart_dataset.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

# Explanation of the Data according to the kaggle

1.	age - age in years
2.	sex - (1 = male; 0 = female)
3.	cp - chest pain type
•	0: Typical angina: chest pain related decrease blood supply to the heart
•	1: Atypical angina: chest pain not related to heart
•	2: Non-anginal pain: typically esophageal spasms (non heart related)
•	3: Asymptomatic: chest pain not showing signs of disease
4.	trestbps - resting blood pressure (in mm Hg on admission to the hospital) anything above 130-140 is typically cause for concern
5.	chol - serum cholestoral in mg/dl
•	serum = LDL + HDL + .2 * triglycerides
•	above 200 is cause for concern
6.	fbs - (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)
•	'>126' mg/dL signals diabetes
7.	restecg - resting electrocardiographic results
•	0: Nothing to note
•	1: ST-T Wave abnormality
	can range from mild symptoms to severe problems
	signals non-normal heart beat
•	2: Possible or definite left ventricular hypertrophy
	Enlarged heart's main pumping chamber
8.	thalach - maximum heart rate achieved
9.	exang - exercise induced angina (1 = yes; 0 = no)
10.	oldpeak - ST depression induced by exercise relative to rest looks at stress of heart during excercise unhealthy heart will stress more
11.	slope - the slope of the peak exercise ST segment
•	0: Upsloping: better heart rate with excercise (uncommon)
•	1: Flatsloping: minimal change (typical healthy heart)
•	2: Downslopins: signs of unhealthy heart
12.	ca - number of major vessels (0-3) colored by flourosopy
•	colored vessel means the doctor can see the blood passing through
•	the more blood movement the better (no clots)
13.	thal - thalium stress result
•	1,3: normal
•	6: fixed defect: used to be defect but ok now
•	7: reversable defect: no proper blood movement when excercising
14.	target - have disease or not (1=yes, 0=no) (= the predicted attribute)



In [7]:
#to analyse some info about the data

In [8]:
heart_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [9]:
#we will check some statistical measures of the data

In [10]:
heart_dataset.isnull().sum().sum()

0

In [11]:
heart_dataset.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

### now we will check how many people are having the disease and how many people are not having the disease. In this scenario 1 == Defective Heart   0 == Healthy Heart

In [12]:
#In this data set the target columns will represents the person is having the healthy heart or not.
heart_dataset.target.value_counts()

1    165
0    138
Name: target, dtype: int64

In [14]:
#Check for missing values

In [15]:
heart_dataset.isnull().sum().sum()

0

In [17]:
#Means there is no null values in the given data set.

In [20]:
#Splitting the features and the target by train_test_split

In [21]:
X = heart_dataset.drop(columns='target', axis=1)
Y = heart_dataset['target']

In [22]:
X

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  
0        0   0     1  
1        0   0     2  
2        2   0     2  
3        2   0     2  
4        2   0     2  
..     ...  ..   ...  
298      1   0     3  
299      1   0     3  
300      1   2     3  
301      1   1     3  
302      1   1     2  

[303 rows x 13 columns]

In [23]:
Y

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: target, Length: 303, dtype: int64

In [24]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.4, random_state=101)

In [25]:
X_train

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
72    29    1   1       130   204    0        0      202      0      0.0   
123   54    0   2       108   267    0        0      167      0      0.0   
225   70    1   0       145   174    0        1      125      1      2.6   
293   67    1   2       152   212    0        0      150      0      0.8   
226   62    1   1       120   281    0        0      103      0      1.4   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
40    51    0   2       140   308    0        0      142      0      1.5   
75    55    0   1       135   250    0        0      161      0      1.4   
87    46    1   1       101   197    1        1      156      0      0.0   
63    41    1   1       135   203    0        1      132      0      0.0   
11    48    0   2       130   275    0        1      139      0      0.2   

     slope  ca  thal  
72       2   0     2  
123      2   0     2  
225      0   0     3  
293      1   0     3  
226      1   1     3  
..     ...  ..   ...  
40       2   1     2  
75       1   0     2  
87       2   0     3  
63       1   0     1  
11       2   0     2  

[181 rows x 13 columns]

In [26]:
Y_train

72     1
123    1
225    0
293    0
226    0
      ..
40     1
75     1
87     1
63     1
11     1
Name: target, Length: 181, dtype: int64

In [27]:
X_train.shape

(181, 13)

In [28]:
X_test.shape

(122, 13)

In [29]:
X.shape

(303, 13)

In [30]:
#Means this function given 60% to train data and 40% to test data

# Training the Model Using Logistic Regression 

In [31]:
from sklearn.linear_model import LogisticRegression
model1 = LogisticRegression(solver='lbfgs', max_iter=1000)

In [32]:
model1.fit(X_train.values, Y_train.values)

LogisticRegression(max_iter=1000)

## Model Evaluation For Logistic Regression Using Accuracy Score 

In [33]:
# accuracy on training data
from sklearn.metrics import accuracy_score
X_train_prediction = model1.predict(X_train.values)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [34]:
print('Accuracy on Training data using Logistic Regression : ', training_data_accuracy*100)

Accuracy on Training data using Logistic Regression :  83.42541436464089


In [35]:
# accuracy on test data
X_test_prediction = model1.predict(X_test.values)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [36]:
print('Accuracy on Test data using Logistice Regression : ', test_data_accuracy*100)

Accuracy on Test data using Logistice Regression :  81.9672131147541


# Training the Model using K-nearest neighbors

In [37]:
from sklearn.neighbors import KNeighborsClassifier

model2 = KNeighborsClassifier()
model2.fit(X_train, Y_train)

KNeighborsClassifier()

# Model Evaluation For K-nearest neighbors Using Accuracy Score

In [38]:
X_train_prediction2 = model2.predict(X_train)
training_data_accuracy2 = accuracy_score(X_train_prediction2, Y_train)

In [39]:
print('Accuracy on Training data using K-nearest neighbors : ', training_data_accuracy2*100)

Accuracy on Training data using K-nearest neighbors :  73.48066298342542


In [40]:
X_test_prediction2 = model2.predict(X_test)
test_data_accuracy2 = accuracy_score(X_test_prediction2, Y_test)

In [41]:
print('Accuracy on Test data using K-nearest neighbors : ', test_data_accuracy2*100)

Accuracy on Test data using K-nearest neighbors :  63.114754098360656


# Training the Model using Decision Tree Classifier

In [42]:
from sklearn.tree import DecisionTreeClassifier

In [43]:
model3 = DecisionTreeClassifier(random_state=101)
model3.fit(X_train, Y_train)

DecisionTreeClassifier(random_state=101)

# Model Evaluation For Decision Tree Classifier Using Accuracy Score

In [44]:
X_train_prediction3 = model3.predict(X_train)
training_data_accuracy3 = accuracy_score(X_train_prediction3, Y_train)

In [45]:
print('Accuracy on Training data using Decision Tree Classifier : ', training_data_accuracy3*100)

Accuracy on Training data using Decision Tree Classifier :  100.0


In [46]:
X_test_prediction3 = model3.predict(X_test)
test_data_accuracy3 = accuracy_score(X_test_prediction3, Y_test)

In [47]:
print('Accuracy on Test data using Decision Tree Classifier : ', test_data_accuracy3*100)

Accuracy on Test data using Decision Tree Classifier :  81.9672131147541


# Training the Model using Support Vector Machine

In [48]:
from sklearn.svm import SVC


model4 = SVC(kernel='rbf', gamma=0.1, C=1.0)
model4.fit(X_train, Y_train)

SVC(gamma=0.1)

# Model Evaluation For Support Vector Machine Using Accuracy Score 

In [49]:
X_train_prediction4 = model4.predict(X_train)
training_data_accuracy4 = accuracy_score(X_train_prediction4, Y_train)
print('Accuracy on Training data using Support Vector Machine : ', training_data_accuracy4*100)

Accuracy on Training data using Support Vector Machine :  100.0


In [50]:
X_test_prediction4 = model4.predict(X_test)
test_data_accuracy4 = accuracy_score(X_test_prediction4, Y_test)

In [51]:
print('Accuracy on Test data using Support Vector Machine : ', test_data_accuracy4*100)

Accuracy on Test data using Support Vector Machine :  50.81967213114754


# Comparing the Accuracy Scores of All Models

In [52]:

#Creating Data Frame for the accuracy score of logistic regression.

results_df = pd.DataFrame(data=[["Logistic Regression", training_data_accuracy*100, test_data_accuracy*100]], 
                          columns=['Model', 'Training Accuracy %', 'Testing Accuracy %'])

#Creating Data Frame for the accuracy score of K-nearest neighbors.

results_df_2 = pd.DataFrame(data=[["K-nearest neighbors",training_data_accuracy2*100 , test_data_accuracy2*100]], 
                          columns=['Model', 'Training Accuracy %', 'Testing Accuracy %'])
results_df = results_df.append(results_df_2, ignore_index=True)

#Creating Data Frame for the accuracy score of Decision Tree Classifier.

results_df_3 = pd.DataFrame(data=[["Decision Tree Classifier",training_data_accuracy3*100, test_data_accuracy3*100]], 
                          columns=['Model', 'Training Accuracy %', 'Testing Accuracy %'])
results_df = results_df.append(results_df_3, ignore_index=True)

#Creating Data Frame for the accuracy score of Support Vector Machine.

results_df_4 = pd.DataFrame(data=[["Support Vector Machine",training_data_accuracy4*100, test_data_accuracy4*100]], 
                          columns=['Model', 'Training Accuracy %', 'Testing Accuracy %'])
results_df = results_df.append(results_df_4, ignore_index=True)

#The Final Data Frame of the accuracy score of all results.

results_df

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_11616\1575563002.py:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results_df = results_df.append(results_df_2, ignore_index=True)
C:\Users\LENOVO\AppData\Local\Temp\ipykernel_11616\1575563002.py:16: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results_df = results_df.append(results_df_3, ignore_index=True)
C:\Users\LENOVO\AppData\Local\Temp\ipykernel_11616\1575563002.py:22: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results_df = results_df.append(results_df_4, ignore_index=True)


Model  Training Accuracy %  Testing Accuracy %
0       Logistic Regression            83.425414           81.967213
1       K-nearest neighbors            73.480663           63.114754
2  Decision Tree Classifier           100.000000           81.967213
3    Support Vector Machine           100.000000           50.819672

## So by Analysing the above table We can easily guess that Logistic Regression is Performing a well Training and Testing accuracy Score. So we will Continue Predicting the new Data with Logistic Regression Model. 

In [53]:
# values of a person heart scan
input_data = (62,0,0,140,268,0,0,160,0,3.6,0,2,2)

# change the input data to a numpy array
input_data_as_numpy_array= np.asarray(input_data)

# reshape the numpy array as we are predicting for only on instance
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

#predicting data using Logistic Regression
prediction = model1.predict(input_data_reshaped)

if (prediction[0]== 0):
  print('According to the given details person does not have a Heart Disease')
else:
  print('According to the given details person has Heart Disease')

According to the given details person does not have a Heart Disease
